lgc/state/PalMetadata.cpp

```diff
@@ -719,94 +719,45 @@ void PalMetadata::eraseColorExportInfo() {
+}
 
-  // Find the first hardware shader stage that has SPI registers set. That must be the API VS, or the merged
-  // shader containing it. User data register 0 is always set if a shader stage is active.
-  static const std::pair<unsigned, unsigned> shaderTable[] = {{mmSPI_SHADER_USER_DATA_LS_0, CallingConv::AMDGPU_LS},
-                                                              {mmSPI_SHADER_USER_DATA_HS_0, CallingConv::AMDGPU_HS},
-                                                              {mmSPI_SHADER_USER_DATA_ES_0, CallingConv::AMDGPU_ES},
-                                                              {mmSPI_SHADER_USER_DATA_GS_0, CallingConv::AMDGPU_GS},
-                                                              {mmSPI_SHADER_USER_DATA_VS_0, CallingConv::AMDGPU_VS}};
-  ArrayRef<std::pair<unsigned, unsigned>> shaderEntries = shaderTable;
-  auto userDataRegIt = m_registers.begin();
-  for (;;) {
-    userDataRegIt = m_registers.find(m_document->getNode(shaderEntries[0].first));
-    if (userDataRegIt != m_registers.end())
-      break;
-    shaderEntries = shaderEntries.drop_front();
-  }
-  unsigned userDataOffset = 0;
-  unsigned userDataReg0 = shaderEntries[0].first;
-  regInfo.callingConv = shaderEntries[0].second;
-  if (userDataReg0 != mmSPI_SHADER_USER_DATA_VS_0 && m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9)
-    userDataOffset = 8; // GFX9+ merged shader: extra 8 SGPRs beore user data
-
-  // Scan the user data registers for vertex buffer table, vertex id, instance id.
-  unsigned userDataCount = m_pipelineState->getTargetInfo().getGpuProperty().maxUserDataCount;
-  for (;;) {
-    switch (static_cast<UserDataMapping>(userDataRegIt->second.getUInt())) {
-    case UserDataMapping::VertexBufferTable:
-      regInfo.vertexBufferTable = userDataRegIt->first.getUInt() - userDataReg0 + userDataOffset;
-      break;
-    case UserDataMapping::BaseVertex:
-      regInfo.baseVertex = userDataRegIt->first.getUInt() - userDataReg0 + userDataOffset;
-      break;
-    case UserDataMapping::BaseInstance:
-      regInfo.baseInstance = userDataRegIt->first.getUInt() - userDataReg0 + userDataOffset;
-      break;
-    default:
-      break;
-    }
-    if (++userDataRegIt == m_registers.end() || userDataRegIt->first.getUInt() >= userDataReg0 + userDataCount)
-      break;
-  }
+// =====================================================================================================================
+// Returns the number of user data registers are used for the shader with the given calling convention.
+//
+// @param callingConv : The calling convention of the shader we are interested in.
+unsigned PalMetadata::getUserDataCount(unsigned int callingConv) {
+  // Should we just define mmSPI_SHADER_PGM_RSRC2_* for all shader types? The comment before the definition of
+  // mmSPI_SHADER_PGM_RSRC2_VS in AbiMetadata.h says that do not want to.
+  static const std::map<unsigned, unsigned> callingConvToRsrc2Map = {
```

Comment:
Do need these to be ordered? If not `llvm::SmallDenseMap` should perform better.

Reply:
Also elsewhere

Reply:
I looked into it, and I'm not sure how to cleanly initialize the value.  It is static, so I only want to do it once.  The initializer list does not work.  I will change it if there is a good way to do that.

Reply:
Ah, that's odd that an initializer list constructor is missing. Can we use a simple helper function instead? Something like:
```
SmallDenseMap<unsigned, unsigned> build(std::initializer_list<std::pair<unsigned, unsigned>> init) {
  return {init.begin(), init.end()};
}
...
static SmallDenseMap m = build({{0, 1}}, {2, 3}});
```

And I will see if we can get a proper constructor added to LLVM and clean this up later.

Reply:
I'll just leave it like this for now.  We can change it later if we want.

Reply:
I added an initializer list constructor in https://github.com/llvm/llvm-project/commit/3c3165cfa0a4ff4f6c687b4ab4c141803d37009d

Reply:
You are right.  The map is not  trivially destructible.  I'll have to change that.

Reply:
I've replace them with an array of pairs, with is POD.